src/mail/view/MailViewModel.ts

```diff
@@ -0,0 +1,351 @@
+import { isOfflineError } from "../../api/common/utils/ErrorCheckUtils.js"
+import { MailFolderType } from "../../api/common/TutanotaConstants.js"
+import { assertSystemFolderOfType, isOfTypeOrSubfolderOf, isSpamOrTrashFolder, isSubfolderOfType } from "../../api/common/mail/CommonMailUtils.js"
+import { mapWith, mapWithout } from "@tutao/tutanota-utils/dist/CollectionUtils.js"
+import { WsConnectionState } from "../../api/main/WorkerClient.js"
+import { WebsocketConnectivityModel } from "../../misc/WebsocketConnectivityModel.js"
+import { ExposedCacheStorage } from "../../api/worker/rest/DefaultEntityRestCache.js"
+import { PreconditionFailedError } from "../../api/common/error/RestError.js"
+import { UserError } from "../../api/main/UserError.js"
+import { ProgrammingError } from "../../api/common/error/ProgrammingError.js"
+
+export enum MailFilterType {
+	Unread,
+	Read,
+	WithAttachments,
+}
+
+export class MailViewModel {
+	private _listId: Id | null = null
+	private _targetMailId: Id | null = null
+	private conversationViewModel: ConversationViewModel | null = null
+	private _filterType: MailFilterType | null = null
+
+	/**
+	 * We remember the last URL used for each folder so if we switch between folders we can keep the selected mail.
+	 * There's a similar (but different) hacky mechanism where we store last URL but per each top-level view: navButtonRoutes. This one is per folder.
+	 */
+	private mailListToSelectedMail: ReadonlyMap<Id, Id> = new Map()
+
+	constructor(
+		private readonly mailModel: MailModel,
+		private readonly entityClient: EntityClient,
+		private readonly eventController: EventController,
+		private readonly connectivityModel: WebsocketConnectivityModel,
+		private readonly cacheStorage: ExposedCacheStorage,
+	) {}
+
+	/** id of the mail that we are trying to load based on the URL */
+	get targetMailId(): Id | null {
+		return this._targetMailId
+	}
+
+	get filterType(): MailFilterType | null {
+		return this._filterType
+	}
+
+	setFilter(filter: MailFilterType | null) {
+		this._filterType = filter
+		this.listModel?.setFilter(this.getFilterForType(filter))
+	}
+
+	private getFilterForType(filter: MailFilterType | null): ListFilter<Mail> | null {
+		switch (filter) {
+			case MailFilterType.Read:
+				return (mail) => !mail.unread
+			case MailFilterType.Unread:
+				return (mail) => mail.unread
+			case MailFilterType.WithAttachments:
+				return (mail) => mail.attachments.length > 0
+			case null:
+				return null
+		}
+	}
+
+	async showMail(listId?: Id, mailId?: Id) {
+		let listIdToUse
+		if (typeof listId === "string") {
+			const mailboxDetail = await this.mailModel.getMailboxDetailsForMailListId(listId)
+			if (mailboxDetail) {
+				listIdToUse = listId
+			} else {
+				listIdToUse = await this.getListIdForUserInbox()
+			}
+		} else {
+			listIdToUse = await this.getListIdForUserInbox()
+		}
+
+		await this.setListId(listIdToUse)
+
+		if (typeof mailId === "string") {
```

Comment:
null check?

Reply:
we are being more strict here and not trusting URL things